main: Tweak runtime GC params for Go 1.19.

davecgh opened the pull request:
Go 1.19 introduced the ability to specify a soft upper memory limit to help deal with transient memory spikes without needing to set the GC percent to an artificially low value.

This takes advantage of that to impose an upper memory limit that leaves plenty of headroom for the minimum recommended value while taking into account the max utxo cache size configuration option and increase the GC percent to the default value of 100% which allows the mem usage to expand more in between each GC cycle which in turn significantly reduces the number of GC cycles that need to be performed, particularly while performing the initial chain sync.

The end result is much less CPU time spent doing garbage collection and thus reduces the amount of time it takes to perform the initial chain sync by about 10%.

The update also has the nice side benefit that the `GOGC` environment variable can be used if an advanced sysadmin really wanted to tune it since it is no longer being overridden.

Finally, it also removes the code that explicitly configures the runtime to use all processor cores because that has been the default since Go 1.5.

The following graphs show OS-reported CPU and memory usage for an initial sync of the first 500k blocks for the existing code vs the updates in this PR.  As can be seen, there is about a 10% speedup in exchange for using more of the available memory:

![dcrd_initial_sync_os_cpu_usage_comparison_500k_blocks_mem_limit_vs_1_8_0_ab226e09](https://user-images.githubusercontent.com/2115102/202367604-95a4df46-dbe9-43d8-8c35-3a7b1da5e2e2.svg)

![dcrd_initial_sync_os_mem_usage_comparison_500k_blocks_mem_limit_vs_1_8_0_ab226e09](https://user-images.githubusercontent.com/2115102/202367623-aa725faf-8a35-45d3-8377-05e0eaa7fb78.svg)

matthawkins90 commented:
I'm trying to follow along with what it means for the GC to be set at 200%.

dcrd's minimum recommended specs suggest 2 GB RAM.

`debug.SetMemoryLimit` sets the memory limit to 1.5 GiB. Is that the soft upper limit or a hard upper limit?

Does the GC being set to 200% mean that the GC will trigger at 3 GiB, or 0.75 GiB?

davecgh commented:
> Does the GC being set to 200% mean that the GC will trigger at 3 GiB, or 0.75 GiB?

Neither.  It means that a collection will be triggered when the ratio of freshly allocated data to live data remaining after the previous collection reaches 200%.

The runtime will try to respect the 1.5GiB memory limit regardless by doing things such as adjusting to the frequency of garbage collections and returning memory to the underlying system more aggressively.

Take a look at the [new garbage collection guide](https://go.dev/doc/gc-guide#Memory_limit) if you want to dig in further and play around with some sliders to see the effects.

matthawkins90 commented:
Thank you! And 10% faster IBD is really impressive!

matthawkins90 commented:
> The SetMemoryLimit function available in the runtime/debug package sets a maximum on the total amount of memory that the Go runtime can use.

Am I correct in understanding that hardcoding a limit of 1.5 GiB is effectively hamstringing systems that have gobs more RAM than 2 GiB available?

Obviously, I agree with catering to tiny systems first so that you don't NEED a powerhouse to run dcrd, but theoretically wouldn't it be better to use a .config variable, or even some sort of `GetSystemMemory` function and then multiplying that by 0.75?

Although, a quick Google search seems to indicate that it's not trivial to get the total system memory in Go.

davecgh commented:
Theoretically, if it actually needed more memory that was available and then artificially limited it, yes.  I did a bunch of profiling though before arriving at these numbers and, in practice, dcrd never actually goes above 1.2 GiB usage during the initial sync.  I intentionally gave it a 25% buffer while still leaving 500 MiB for runtime overhead that is used external to Go such as OS kernel memory held on behalf of the process.

I suppose I should've called it out already, but I do plan to add a config option once we require Go 1.19 as a minimum version.  I didn't do it here because of the aforementioned profiling and because there isn't really an easy way to make it only have the CLI option depending on the Go version building it (it's not impossible, but it would require a lot of extra work that doesn't seem worth it only to have to remove it once we require Go 1.19 at the next Go release anyway).

matthawkins90 commented:
Sounds good. Thanks as usual for the great explanations!

matheusd commented:
Plots for memory (RSS) and cpu usage from before and after this commit, compiled with go 1.19.3

This is syncing up to about height 600k. Difference in time was about 5% less (with this commit on). 


![plot-pre-gc](https://user-images.githubusercontent.com/1752493/201990205-941ab151-ca63-43a2-b22c-ca3ee9df8f40.svg)
![plot-pos-gc](https://user-images.githubusercontent.com/1752493/201990221-19d84326-c891-4667-99af-53b4320270ce.svg)

davecgh commented:
As I mentioned in the dev channel, my original testing on this was done when the new Go 1.19 feature was still in dev and they've made some further tweaks since then.

I went ahead and tested this again without increasing the GC percent to 200% and allowed it to remain at the default of 100% and there is very little difference in terms of performance and it actually ends up doing slightly fewer GC cycles because it only reaches a peak of 1.2 GiB memory usage with a GC percent of 100% versus requiring a few extra forced GCs due to running into the 1.5 GiB limit with a GC percent of 200%.

I also realized that this needs to account for the fact the utxo cache size is configurable and can be raised and thus blindly setting the limit to 1.5 GiB without taking that into account would make the option not work as intended.

Thus, I've reworked this a bit to leave the GC percentage at the default of 100% starting with Go 1.19 and to increase the soft memory limit if needed depending on the `--utxocachemaxsize` option.  I also added a humanized log output for the configured memory limit when enabled.

The updates mean that it is now possible to increase the limit per the previous discussion via the `--utxocachemaxsize` option and also has the nice side benefit that the `GOGC` environment variable can be used if an advanced sysadmin really wanted to tune it since it is no longer being overridden.

davecgh commented on dcrd.go:
```diff
@@ -73,6 +73,37 @@ func dcrdMain() error {
+	// Starting with Go 1.19, a soft upper memory limit is imposed that leaves
+	// plenty of headroom for the minimum recommended value and the target GC
+	// percentage is left at the default value to significantly reduce the
+	// number of GC cycles thereby reducing the amount of CPU time spend doing
```
Updated.

davecgh commented:
The following charts show the difference between the 100% and 200% GC Percent along with a 1.5 GiB memory limit.  As can be seen, the CPU usage is about the same, but the 200% variant reaches the 1.5 GiB limit and stays there while the 100% variant slowly raises until it peaks out at 1.2 GiB:

![dcrd_initial_sync_gc100_vs_200_os_cpu_usage_comparison_500k_blocks](https://user-images.githubusercontent.com/2115102/202352314-cce8c6f3-598f-4434-8544-d9860222d253.svg)

![dcrd_initial_sync_gc100_vs_200_os_mem_usage_comparison_500k_blocks](https://user-images.githubusercontent.com/2115102/202352328-7bf94913-f115-4e5b-b999-eb293caa3451.svg)

davecgh commented:
I modified the PR description to include charts comparing the OS-reported CPU and memory usage of of the current code vs this PR.

matheusd reviewed (approved):
Tested with the new behavior (including changing the utxo cache size)
